serialisable.hpp

```diff
@@ -347,6 +378,48 @@ class Serialisable {
 		return true;
 	}
 
+	/*!
+	* \brief Saves or loads an arithmetic integer value
+	* \param The name of the value in the output/input file
+	* \param Reference to the value
+	* \return false if the value was absent while reading, true otherwise
+	*/
+	template<typename T>
+	typename std::enable_if<std::is_integral<T>::value, bool>::type
+		synch(const std::string& key, T& value) {
+		if (preferencesSaving_) {
+			preferencesJson_->getObject()[key] = std::make_shared<JSONint>(int64_t(value));
+		}
+		else {
+			auto found = preferencesJson_->getObject().find(key);
+			if (found != preferencesJson_->getObject().end()) {
+				value = T(found->second->getInt());
+			} return false;
+		}
+		return true;
+	}
+
+	/*!
+	* \brief Saves or loads an enum as integer
+	* \param The name of the value in the output/input file
+	* \param Reference to the value
+	* \return false if the value was absent while reading, true otherwise
+	*/
+	template<typename T>
+	typename std::enable_if<std::is_enum<T>::value, bool>::type
```

Comment:
`enum` values are serialised as integers. Implicit conversion between integers and `enum`s does the trick.

Reply:
I think there is a little problem here. It may happen that the JSON file was written by another JSON library, it may be a float that does not contain a dot. This would cause exceptions when expecting a float and the number is identified as an integer.

It can be easily solved by reimplementing the `getDouble()` method in the `JSONint` class as well.

Reply:
I thought about this, but wondered if every `JSONobject` was supposed to only implement one method. Glad to see I was partly right. I will fix it in the next commit, thank you for pointing out.

Reply:
There is no rule that every `JSONobject` descendant is supposed to implement only one method, it just didn't make sense to implement more than one so far.

Reply:
The fix you proposed may be trickier than foreseen - binding `double&` to unrelated type `int` is not supported (at least on `MSVC2015`). The solution I found so far is to define two `toDouble()` methods: a `double& JSONobject::toDouble()` and a `double JSONobject::toDouble() const`. `JSONdouble` will override both, and the second one can be overridden by `JSONint` too. Not completely satisfied with this though. Will commit the other fix and wait for your feedback on this.

For reference:
```
struct A {
	double& D1() { throw 1; }
	double D1() const { throw 2; }

};

struct D {
	double v1;
	int v2;

	double D1() const { return v2; }

};

int main()
{
	D d;
	d.v1 = 1;
	d.v2 = 2;
	double db = d.D1();
	return 0;
}
```

Reply:
I have forgotten there's that unpleasant reference. That complicates things a lot. That trick with const method would not be so easy to apply, because there is a `const_cast` and the serialisation is not a const method. It could be _un_-consted in the `synch()` methods instead, but that would possibly lead to a lot of extra code, requiring all JSON getters to have both a non-const and a const method.

An easier solution would be to create some `JSONnumber` class with a private constructor (with the serialisation stuff as friend) that would be created only by the parser when it's in doubt and contain both `double` and `long int`.

Reply:
Sure, go ahead. I'll have a look at the solution with pleasure once it's done!